### PR TITLE
Add a torch version of DDPG

### PR DESCRIPTION
**Pull request by ShuaibinLi**
## A torch/parl version of ddpg

add files to :
ddpg/mujoco_model.py,mujoco_agent.py,train.py,README.md ---> PARL/benchmark/torch/
ddpg.py ---> PARL/parlalgotithms/torch/

### CODE REVIEW

File: benchmark/torch/ddpg/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+Based on PARL, the DDPG algorithm of deep reinforcement learning has been reproduced, reaching the same level of indicators as the paper in Mujoco benchmarks.
+
+> Paper: DDPG in [Continuous control with deep reinforcement learning](https://arxiv.org/abs/1509.02971)
+
+### Mujoco games introduction
+Please see [here](https://github.com/openai/mujoco-py) to know more about Mujoco games.
+
+### Benchmark result
+
+<img src=".benchmark/evaluation.png" width = "800" height ="300" alt="DDPG_HalfCheetah-v2"/> <img src=".benchmark/DDPG_Humanoid-v2.png" width = "400" height ="300" alt="DDPG_HalfCheetah-v1"/>
```
Comment (rical730) [inline benchmark/torch/ddpg/README.md]: 2 pictures ?

File: benchmark/torch/ddpg/mujoco_agent.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+        self.device = torch.device("cuda" if torch.cuda.
+                                   is_available() else "cpu")
+
+        self.alg.sync_target(decay=0)
+
+    def predict(self, obs):
+        obs = torch.FloatTensor(obs.reshape(1, -1)).to(self.device)
+        action = self.alg.predict(obs)
+        return action.cpu().data.numpy().flatten()
```
Comment (rical730) [inline benchmark/torch/ddpg/mujoco_agent.py]:
```
action_numpy = action.cpu().data.numpy().flatten()
return action_numpy
```

File: parl/algorithms/torch/ddpg.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import parl
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+'''
+Algorithm of DDPG: describes the mechanism of updating parameters in Model
+
+Contributions: predict(obs): return act         # choose action based on actor net
+               learn(replay_memory_batch)       # 
+               sync_target()                    # update networks
+'''
```
Comment (rical730) [inline parl/algorithms/torch/ddpg.py]: please remove these comments, we have API document.

File: benchmark/torch/ddpg/train.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+        # Train agent after collecting sufficient data
+        if t >= args.start_timesteps:
+            batch_obs, batch_action, batch_reward, batch_next_obs, batch_terminal = replay_memory.sample_batch(
+                args.batch_size)
+            agent.learn(batch_obs, batch_action, batch_reward, batch_next_obs,
+                        batch_terminal)
+
+        if done:
+            # +1 to account for 0 indexing. +0 on ep_timesteps since it will increment +1 even if done=True
```
Comment (rical730) [inline benchmark/torch/ddpg/train.py]: You can remove this comment since the log print is clear enough

**Review by rical730 — COMMENTED** [review]
please register ddpg in  parl/algorithms/torch/__init__.py and make sure you can run train.py after install parl

File: benchmark/torch/ddpg/mujoco_agent.py (hunk 2 above)
Comment (ShuaibinLi) [reply benchmark/torch/ddpg/mujoco_agent.py]: image

Resolved.

File: benchmark/torch/ddpg/train.py (hunk 4 above)
Comment (ShuaibinLi) [reply benchmark/torch/ddpg/train.py]: image
Resolved.

File: benchmark/torch/ddpg/train.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        state, done = eval_env.reset(), False
+        while not done:
+            action = agent.predict(np.array(state))
+            state, reward, done, _ = eval_env.step(action)
+            avg_reward += reward
+    avg_reward /= eval_episodes
+    return avg_reward
+
+
+if __name__ == "__main__":
```
Comment (rical730) [inline benchmark/torch/ddpg/train.py]: please split this part to several functions

Comment (ShuaibinLi) [reply benchmark/torch/ddpg/train.py]: Resolved. Have split it to three defs(main, run_train_episode, run_evaluate_episodes)
Also rerun it on GPU server, the performance has not decreased and even much better for seed 0.

image
image
image

File: benchmark/torch/ddpg/train.py [hunk 6]
```diff
@@ -23,9 +23,47 @@
 from parl.algorithms import DDPG
 
 
+# Run episode for training
+def run_train_episode(agent, env, replay_memory, args):
```
Comment (rical730) [inline benchmark/torch/ddpg/train.py:27]: `replay_menory` --> `rpm`
args is a global variable

File: benchmark/torch/ddpg/train.py [hunk 7]
```diff
@@ -23,9 +23,47 @@
...
+    while not done:
+
+        episode_steps += 1
+        # Select action randomly or according to policy
+        if replay_memory.size() < args.start_timesteps:
+            action = env.action_space.sample()
+        else:
+            action = (agent.predict(np.array(obs)) + np.random.normal(
+                0, max_action * args.expl_noise, size=action_dim)).clip(
+                    -max_action, max_action)
```
Comment (rical730) [inline benchmark/torch/ddpg/train.py:41]: this line is poorly readable

File: benchmark/torch/ddpg/train.py [hunk 8]
```diff
@@ -101,57 +106,66 @@ def eval_policy(agent, env_name, seed, eval_episodes=5):
     replay_memory = ReplayMemory(
```
Comment (rical730) [inline benchmark/torch/ddpg/train.py:106]: `replay_memory ` --> `rpm`

**Pull request closed without merging**